### PR TITLE
Admin interface.

### PR DESCRIPTION
**Pull request by zhzhzoo**

### CODE REVIEW

File: application/controllers/admin.php [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+
+    private function is_logged_in() {
+        return $this->session->userdata('logged_in') === true;
+    }
+
+    private function check_permission() {
+        if ($this->is_logged_in()) {
+            return;
+        }
+        redirect('/admin/auth/');
```
Comment (JeldorPKU) [inline application/controllers/admin.php]: 这里可以`site_url()`来弄会不会比较好？

File: application/controllers/admin.php [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+            //都对了
+            $this->session->set_userdata('logged_in', true);
+            redirect('/admin/ls/');
+        }
+
+        show_404('');
+    }
+
+	public function index() {
+        redirect('/admin/auth/', 'location', 301);
```
Comment (JeldorPKU) [inline application/controllers/admin.php]: 嗯这里也是，用`site_url()`吧。

File: application/config/constants.php [hunk 3]
```diff
@@ -36,16 +36,7 @@
...
- * ------------------------------------------------------------------------
- * Province Settings
- * ------------------------------------------------------------------------
- *
- * This array is for the dropdown-menu in the sign up form.
- *
- */
-
-$PROVINCES = array(
+$GLOBALS['PROVINCES'] = array(
```
Comment (JeldorPKU) [inline application/config/constants.php]: 貌似这里不用这么写？

Comment (JeldorPKU) [inline application/config/constants.php]: 另外不要删注释好吧……

File: application/models/people_model.php [hunk 4]
```diff
@@ -10,28 +10,46 @@ class People_model extends CI_Model {
...
+        $res = $this->db->select('people.*, users.school')->from('people')->where('deleted', false)->where($where)->join('users', 'people.school_id = users.id') -> get();
+        return $res;
+    }
+
+    /*
+     * Get a person's information.
+     */
     public function get_people($id) {
-        $res = $this->db->where('deleted', false)->where('id', $id)->get('people');
+        $res = $this->get_people_where(array('people.id' => $id));
```
Comment (JeldorPKU) [inline application/models/people_model.php]: 人员信息可能被删除了，要不要考虑一下？

File: application/controllers/admin.php (hunk 1 above)
Comment (zhzhzoo) [inline application/controllers/admin.php]: 嗯手册表示 "Does a "header redirect" to the URI specified. If you specify the full site URL that link will be build, but for local links simply providing the URI segments to the controller you want to direct to will create the link. The function will build the URL based on your config file values."
然后 ｀system/helpers/url_helper.php｀ 里是这么写的

        if ( ! preg_match('#^https?://#i', $uri))
        {
            $uri = site_url($uri);
        }

File: application/config/constants.php (hunk 3 above)
Comment (zhzhzoo) [inline application/config/constants.php]: 啊不行啊 $PROVINCES 被 include 以后在外面那个文件里就找不到了。。。这就是为什么前面都是定义的常量的原因，但是 php 5.(5还是6来着) 前的版本都没有常量数组。。。

File: application/models/people_model.php (hunk 4 above)
Comment (zhzhzoo) [inline application/models/people_model.php]: 啊这个考虑了，get_people_where 得到的是没有删除过的~

File: application/config/constants.php (hunk 3 above)
Comment (zhzhzoo) [inline application/config/constants.php]: 唔 抱歉之前四处 C-c C-v 来着把注释弄掉了(･･;) 现在去粘回来

File: application/controllers/admin.php (hunk 1 above)
Comment (huxuan) [inline application/controllers/admin.php]: 我建议按照 @JeldorPKU 的意见把使用 `site_url()` 作为一项 convention.

File: application/config/constants.php [hunk 5]
```diff
@@ -44,8 +44,7 @@
  * This array is for the dropdown-menu in the sign up form.
  *
  */
-
```
Comment (huxuan) [inline application/config/constants.php]: Please revert this change as it's necessary to have an extra line between comment and constant. Just to keep it the same with others.

File: application/controllers/admin.php [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+                    'type' => 'boolean'
+                ),
+                'meal' => array(
+                    'type' => 'boolean'
+                ),
+                'race' => array(
+                    'type' => 'enum',
+                    'enum' => $GLOBALS['RACES']
+                )//,
+                //'team_id' => array(
```
Comment (huxuan) [inline application/controllers/admin.php]: Are you sure to make these as comments?

File: application/controllers/admin.php [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+                'group_id' => array(
+                    'type' => 'foreignkey',
+                    'references' => 'group',
+                    'display_column' => 'group_name'
+                )
+            )
+        );
+    }
+
+    private function get_captcha() {
```
Comment (huxuan) [inline application/controllers/admin.php]: I wonder whether we can make this as a GLOBAL helper as we may need it in other place.

File: application/models/people_model.php (hunk 4 above)
Comment (huxuan) [inline application/models/people_model.php]: 这种写法挺好的，因为deleted的数据一般都是没有意义的（但是没必要删），所以没有必要把`->where('deleted', false)`写的到处都是（代码片段重复多了难免会漏了或者出错），应该是在比较底层的接口筛掉了，然后其他上层的接口都来调用这个底层接口。

**Comment by huxuan** [comment]
Hi @zhzhzoo, I really appreciate what you have done! Just some small note.
Another suggestion it that It will be better if you can split this Pull Request into several ones.

File: application/controllers/admin.php (hunk 1 above)
Comment (zhzhzoo) [inline application/controllers/admin.php]: 啊我觉得所有能不用 `site_url` 的地方都不用 `site_url` 比较好，因为所有自动补全 url 的地方都是用 `site_url` 补的~ 自己写一遍有点繁琐。。。

File: application/config/constants.php (hunk 5 above)
Comment (zhzhzoo) [inline application/config/constants.php]: Ah sorry didn't take care, I'm reverting this~

File: application/controllers/admin.php (hunk 6 above)
Comment (zhzhzoo) [inline application/controllers/admin.php]: Ah it was just I didn't know how to join another table multiple times or how to handle cross-referenced foreign key. I'll uncomment them after solving these.

File: application/controllers/admin.php (hunk 7 above)
Comment (zhzhzoo) [inline application/controllers/admin.php]: Actually I think this should go into model.

**Comment by zhzhzoo** [comment]
er... the problem is that it won't run if I split this Pull Request into several ones....
